---
title: "TFM"
author: "Ainhoa Garcia"
date: "7/4/2021"
output: html_document
---

```{r}
# Load required packages 
library(shiny)
library(shinydashboard)
library(tidyverse)
library(ggiraph)
library(DT)
library(plotly)
```



```{r}
circ <- as.data.frame(read.table("C:/Users/ainho/eccDNA/BED/SRR6315424_unknown_circle.bed",header = FALSE, sep="\t",stringsAsFactors=FALSE))

names(circ) <- c("chrom","start","end","discordant_reads","split_reads","score","coverage_mean","coverage_sd","coverage_start", "coverage_end","coverage_cont")

# Add quality levels (score < 10 = Bad, score < 50 = Low, score < 200 = Medium, score > 200 = Good)
    circ$quality <- cut(circ$score,breaks=c(-Inf,10,50,200,Inf),labels= c("Bad","Low", "Medium", "Good"),right = FALSE)
    
# Set chr as factor
circ$chr <- substr(circ$chr, start = 1, stop = 5)
circ$chr<-str_remove(circ$chr,"_")
circ$chr<-str_remove(circ$chr,"chr")
circ$chr <- factor(circ$chr, levels = c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","Un","M","X","Y"))

# Add circle id
circ <- circ %>% mutate(circle_id = 1:n()) %>% select(circle_id, everything())

# Add size of circle (number of bp)
size_bp <- circ$end - circ$start
circ<- add_column(circ,size_bp)
head(circ)

```

```{r}
circ %>% group_by (quality) %>% summarise((count = n() / nrow(.))*100)
circ %>% group_by (quality) %>% count() %>% mutate(perc = n / nrow(circ)) -> circ_quality
boxplot(circ$size_bp)
```
`

```{r}
total_circles<-circ %>% count()
p_bad <-((circ %>% filter (quality=="bad") %>% count())/total_circles)*100
p_low <- ((circ %>% filter(quality=="low") %>% count())/total_circles)*100
p_medium <- ((circ %>% filter(quality=="medium") %>% count())/total_circles)*100
p_good <- ((circ %>% filter(quality=="good") %>% count())/total_circles)*100
```

```{r}
circ %>% select(chr)%>% distinct() %>% arrange(chr)
```

```{r}
###### EXAMPLE - Submit button changed tab
library(shiny)
ui <- fluidPage(
        tabsetPanel(
                id="inTabset",
                tabPanel("Tab 1",actionButton("switch_tab", "Go to the third tab")
                ),
                tabPanel("Tab 2", "there!"),
                tabPanel("Tab 3", "there!"))

)

server <- function(input, output, session) {

        observeEvent(input$switch_tab, {
                updateTabsetPanel(session, "inTabset",selected = "Tab 3")
        })

}
shinyApp(ui = ui, server = server)
```


```{r}

ui <- dashboardPage(
  dashboardHeader(title = "Simple tabs"),
  dashboardSidebar(
    sidebarMenu(
      id = "tabs",
      menuItem("Dashboard", tabName = "dashboard", icon = icon("dashboard")),
      menuItem("Widgets", tabName = "widgets", icon = icon("th"))
    ),
    actionButton('switchtab', 'Switch tab')
  ),
  dashboardBody(
    tabItems(
      tabItem(tabName = "dashboard",
        h2("Dashboard tab content")
      ),
      tabItem(tabName = "widgets",
        h2("Widgets tab content")
      )
    )
  )
)

server <- function(input, output, session) {
  observeEvent(input$switchtab, {
    newtab <- switch(input$tabs,
      "dashboard" = "widgets",
      "widgets" = "dashboard"
    )
    updateTabItems(session, "tabs", newtab)
  })
}

shinyApp(ui, server)

```



```{r}
ggplot(filter_circ, aes( x = size_bp, y = coverage_cont,color=discordant_reads))+
    geom_point()
```

```{r}
library(ggplot2)

library(ggiraph)

g <- ggplot(circ, aes( x = size_bp, y = coverage_cont))

my_gg <- g + geom_point_interactive(aes(tooltip = split), size = 2)
ggiraph(code = print(my_gg), hover_css = "cursor:pointer;fill:red;stroke:red;")
```

```{r}
quantile(circ$size_bp,c(0.90))
```

```{r}
circles<-c(7984,10519,7227,6825,9675,9250,430,1917,6111,5517,2726,3157,1337,623,617,512,2896,1058,787,5490,2469,1577,625,350,254,471,601,404,454,508,2479,326,4497,3026,2978,1786,4627,5855)
mean(circles)
```


```{r}
renderPlot({
    ggplot(filter_circ(), aes( x = chr, y = size_bp))+
      geom_point(aes(color=chr),size=4,shape=21)+
      theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))
```

```{r}

# Set chromosome as factor
circ$chr <- substr(circ$chr, start = 1, stop = 5)
circ$chr<-str_remove(circ$chr,"_")
circ$chr<-str_remove(circ$chr,"chr")
circ$chr <- factor(circ$chr, levels = c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","Un","M","X","Y"))

# Add quality levels (score < 10 = Bad, score < 50 = Low, score < 200 = Medium, score > 200 = Good)
circ$quality <- cut(circ$score,breaks=c(-Inf,10,50,200,Inf),labels= c("Bad","Low", "Medium", "Good"),right = FALSE)

# Get overall results
total_circles<-circ %>% count()
p_bad <-((circ %>% filter (quality=="Bad") %>% count())/total_circles)*100
p_low <- ((circ %>% filter(quality=="Low") %>% count())/total_circles)*100
p_medium <- ((circ %>% filter(quality=="Medium") %>% count())/total_circles)*100
p_good <- ((circ %>% filter(quality=="Good") %>% count())/total_circles)*100

# Add size of circle (number of bp)
size_bp <- circ$end - circ$start
circ<- add_column(circ,size_bp)
head(circ)

# There are many differences in size between circles, we need to make different graphs
df_circ <- circ %>% filter (size_bp <= 5000)
outliers <- circ %>% filter (size_bp > 5000)


```


```{r}
tabItem(
               tabName = "results",
                    fluidRow(
                        valueBox(total_circles,tags$b("TOTAL DNA CIRCLES"), color="maroon",
                                 icon = tags$i(class = "fas fa-dna", style="font-size: 50px")
                                 ),
                        box(width = 3,background = "olive",tags$h4("GOOD:",paste0(round(p_good,2),"%"))),
                        box(width = 3, background = "yellow",tags$h4("LOW:", paste0(round(p_low,2),"%"))),
                        box(width = 3, background = "light-blue",tags$h4("MEDIUM:",paste0(round(p_medium,2),"%"))),
                        box(width = 3, background = "red",tags$h4("BAD:", paste0(round(p_bad,2),"%"))),
                        ),
                    fluidRow(
                        box(girafeOutput("results"),width=7,
                            status = "primary",
                            ),
                        box(title= span(icon("filter"), "Filters"), status = "warning", width = 5,
```


```{r}
checkboxGroupInput(inputId = "quality",
                                               "Quality of your circles:",
                                               choices = c("Bad","Low","Medium","Good"),
                                               selected= c("Medium","Good")
```


```{r}
),
                            sliderInput(inputId = "size",
                                        "By size (number of base pairs):",
                                        min = min(df_circ$size_bp),
                                        max= 5000,
                                        value = c(250,2500)),
                            
                           
    )
                ),
            )
```

```{r}
filter_circ <- reactive({
    filter_circ<- filter(circ, between(size_bp, input$size[1], input$size[2]), quality == input$quality)
  })
  
  output$results <- renderGirafe({
```


```{r}
gg_scatter <-ggplot(filter_circ(), aes( x = chr, y = size_bp,tooltip=split_reads,data_id=split_reads))+
      geom_point_interactive (size=4,shape=21,color=chr)+
      theme_minimal()
    
    girafe(ggobj = gg_scatter,
           options = list(opts_selection(type = "single")) )
  })
  
```


```{r}
library(Biostrings)
seq <- sample(DNA_ALPHABET[1:4], size=120, replace=TRUE)
pos <- seq(1, length(seq), by=1)
df_seq <- data.frame(pos,seq)
```


```{r}
set.seed(999)
n = 1000
df = data.frame(sectors = sample(letters[1:8], n, replace = TRUE),
    x = rnorm(n), y = runif(n))
library(circlize)
circos.par("track.height" = 0.1)
circos.initialize(df$sectors, x = df$x)
```


```{r}
library(plotly)

      circ <-read.table(input$bedfile$datapath,header = FALSE, sep="\t",stringsAsFactors=FALSE)
      
      
      names(circ) <- c("chr","start","end","discordant_reads","split_reads","score","coverage_mean","coverage_sd","coverage_start", "coverage_end","coverage_cont")
      
      # Set chromosome as factor and fix random chr
      circ$chr <- substr(circ$chr, start = 1, stop = 5)
      circ$chr<-str_remove(circ$chr,"_")
      circ$chr<-str_remove(circ$chr,"chr")
      circ$chr <- factor(circ$chr, levels = c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","Un","M","X","Y"))
      
      # Add quality levels (score < 10 = Bad, score < 50 = Low, score < 200 = Medium, score > 200 = Good)
      circ$quality <- cut(circ$score,breaks=c(-Inf,10,50,200,Inf),labels= c("Bad","Low", "Medium", "Good"),right = FALSE)
      
      # Add circle id
      circ <- circ %>% mutate(circle_id = 1:n()) %>% select(circle_id, everything())
      
      # Add size of circle (number of bp)
      circ$size_bp <- circ$end - circ$start
      circ <- circ %>% filter(size_bp < 5000 & quality!="Bad")

      fig <- circ %>%
  plot_ly(type = 'scatter',
    mode = 'markers',
    x = ~chr,
    y = ~size_bp,
    color = ~quality,
    text = ~circle_id,
    hovertemplate = paste(
      "<b>Circle ID:%{text}<br><br>",
      "%{yaxis.title.text}: %{y:$,.0f}<br>",
      "%{xaxis.title.text}: %{x:.0%}<br>",
      "Number Employed: %{marker.size:,}",
      "<extra></extra>"
      )
    ) 
fig <- fig %>%
  layout(legend = list(orientation = 'h', y = -0.3))

fig
```



```{r}
library(plotly)

fig <- plot_ly(circ,
  type = 'scatter',
  x = ~chr,
  y = ~size_bp,
  text = paste("Discordant Reads: ", circ$discordant_reads,
               "<br>Split Reads : ", circ$split_reads,
               "<br>Size: ", circ$size_bp),
  hoverinfo = 'text',
  mode = 'markers',
  transforms = list(
    list(
      type = 'groupby',
      groups = circ$chr,
      styles = list(
        list(target = 1, value = list(marker =list(color = 'blue'))),
        list(target = 2, value = list(marker =list(color = 'red'))),
        list(target = 3, value = list(marker =list(color = 'black')))
      )
      )
    )
  )

fig
```

```{r}
library(ggiraph)
               gg_results <-ggplot(plot_circ(), aes( x = chr, y = size_bp,color=chr))+
                geom_point_interactive (aes(tooltip=discordant_reads),size=4)+
                 xlab("Original Chromosome")+
                 ylab("Size")+
                 theme_minimal()+
```

```{r}
 
perc<-as.data.frame(circ %>% group_by(size_bp < 5000) %>% count() %>% mutate(pct_tot = n/nrow(circ)*100))
perc_big <- perc[1,3]
perc_smal <- perc[2,3]
```
```{r}
library(dplyr)
mycoords.list<-list("12:15897961:16877384")
mycoords.gr = lapply(mycoords.list, function (x) {res=strsplit(x, ':')}) %>%
   unlist %>%
   as.numeric %>%
   matrix(ncol=3, byrow=T) %>%
   as.data.frame %>%
   select(chrom=V1, start=V2, end=V3) %>%
   mutate(chrom=paste0('chr', chrom)) %>%
   makeGRangesFromDataFrame
mycoords.gr
```


```{r}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(GenomicRanges)
library(tidyverse)
library(org.Hs.eg.db)

circ <- as.data.frame(read.table("C:/Users/ainho/eccDNA/BED/SRR6315423_unknown_circle.bed",header = FALSE, sep="\t",stringsAsFactors=FALSE))

names(circ) <- c("chrom","start","end","discordant_reads","split_reads","score","coverage_mean","coverage_sd","coverage_start", "coverage_end","coverage_cont")
circ$size_bp <- circ$end - circ$start
circ$quality <- cut(circ$score,breaks=c(-Inf,10,50,200,Inf),labels= c("Bad","Low", "Medium", "Good"),right = FALSE)

coords<- circ %>% filter(quality %in% c("Medium","Good") & size_bp < 5000) %>% dplyr::select(chrom,start,end) %>% makeGRangesFromDataFrame 
genes <-genes(TxDb.Hsapiens.UCSC.hg19.knownGene)

genes_df <- as.data.frame(subsetByOverlaps(genes,coords))

entrezid <- genes_df$gene_id

hs <- org.Hs.eg.db
genesymbol_df<- AnnotationDbi::select(hs, 
       keys = entrezid,
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "ENTREZID")

names(genesymbol_df)[names(genesymbol_df) == "ENTREZID"] <- "gene_id"

gene_list<-left_join(genes_df,genesymbol_df)

DT::datatable(gene_list)
```

```{r}
library(RITANdata)
library(RITAN)

my_genes<-gene_list$SYMBOL

resources <- c("ReactomePathways", "KEGG_filtered_canonical_pathways")
e <- term_enrichment(gene_list$SYMBOL, resources = resources, all_symbols = cached_coding_genes)
e$name <-  s1 <- sapply(strsplit(e$name, split='.', fixed=TRUE), function(x) (x[2]))
head(e)

ggplot(e[1:6,],aes(x=p,y=name,fill=p))+
                           geom_col()+
                           scale_color_brewer("RdPu")
                       
```
```{r}
my.dnastring <- as.character(Biostrings::getSeq(BSgenome.Mmusculus.UCSC.mm10, "chr1", 3000000, 3000100))

```

```{r}
onecirc<-circ [1,]

library(biomaRt)
mart <- useMart("ensembl")
mart <- useDataset("hsapiens_gene_ensembl", mart)
attributes <- c("ensembl_gene_id","start_position","end_position","strand","hgnc_symbol","chromosome_name","entrezgene","ucsc","band")
filters <- c("chromosome_name","start","end")
values <- list(chromosome="1",start="1783590",end="2597819")
all.genes <- getBM(attributes=attributes, filters=filters, values=values, mart=mart)
```

```{r}
splitColumnByOverlap <-
    function(query, subject, column="ENTREZID", ...)
{
    olaps <- findOverlaps(query, subject, ...)
    f1 <- factor(subjectHits(olaps),
                 levels=seq_len(subjectLength(olaps)))
    splitAsList(mcols(query)[[column]][queryHits(olaps)], f1)
    }

 symInCnv = splitByOverlap(gns, cnv, "SYMBOL")
```

```{r}
library(ensembldb)
library(EnsDb.Hsapiens.v75)
names(listTables(EnsDb.Hsapiens.v75))
```


```{r}
 # TABLE OUTPUT OF BIG CIRCLES
    table_results<-reactive({
        if (is.null(input$bedfile))
            return(NULL)                
        circ <-read.table(input$bedfile$datapath,header = FALSE, sep="\t",stringsAsFactors=FALSE)
        names(circ) <- c("chr","start","end","discordant_reads","split_reads","score","coverage_mean","coverage_sd","coverage_start", "coverage_end","coverage_cont")
        
        # Add circle id
        circ <- circ %>% mutate(circle_id = 1:n()) %>% select(circle_id, everything())
        
        # Add quality levels (score < 10 = Bad, score < 50 = Low, score < 200 = Medium, score > 200 = Good)
        circ$quality <- cut(circ$score,breaks=c(-Inf,10,50,200,Inf),labels= c("Bad","Low", "Medium", "Good"),right = FALSE)
        
        # Add size of circle (number of bp)
        circ$size_bp <- circ$end - circ$start
        
        # Remove bad coverage circles and outliers with wrong discordant reads outputs
        circ <- circ %>% 
            filter(coverage_cont < 0.5 & size_bp >5000)  %>%
            select (circle_id, chr,start,end,discordant_reads,split_reads,score,quality,size_bp)
        
        # Display only some columns
        names(circ) <- c("Circle ID", "Original Chromosome","Start","End","Discordant Reads","Split Reads","Score","Quality","Size")
        circ
    })
    
    # Make the data table for circles over 5000bp
    output$bigcircles <- DT::renderDataTable({
        DT::datatable(table_results(),options = list(
            searching = FALSE,
            pageLength = 5,
            lengthMenu = c(5, 10, 15, 20),
            scrollX=TRUE,
            columnDefs = list(list(className = 'dt-center', targets = 0:4))
        ))
    })
    
    # Render UI - Appear after click on the first page
    output$table_bigcircles <- renderUI({
        req(input$bedfile)
        box(title="Circles over 5000bp",status="danger",width = NULL, solidHeader = TRUE,
            dataTableOutput("bigcircles"))
    })
    
```

```{r}

```


